{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "# Jovian analysis report\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Please visualize the report by pressing `Cell` in the toolbar and then selecting `Run All`. This can take a couple of minutes (depending on the size of your dataset).**  \n",
    "<br> \n",
    "*N.B. The sum total of reads in this report will not add up to the sum total number of reads that were supplied as input. This is because, 1) human reads are removed, 2) PCR-duplicates might be removed depending on the chosen configuration, by default, PCR-duplicates are not removed.*  \n",
    "<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%html\n",
    "<style>\n",
    "div.output_subarea {\n",
    "    padding-top: 0 !important;\n",
    "    padding-bottom: 0 !important;\n",
    "}\n",
    "\n",
    ".standardheader {\n",
    "    font-size: 30px !important;\n",
    "    padding: 0px;\n",
    "}\n",
    "\n",
    ".standardanchor {\n",
    "    font-size: 14px !important;\n",
    "    margin-left: 5px;\n",
    "}\n",
    ".standardtext {\n",
    "    font-family: sans-serif;\n",
    "    font-size: 14px !important;\n",
    "    margin-left: 5px;\n",
    "}\n",
    "\n",
    ".rendered_html {\n",
    "    margin-left: 5px;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false,
    "tags": []
   },
   "outputs": [],
   "source": [
    "######################################\n",
    "# Required packages for this script  #\n",
    "######################################\n",
    "import pandas as pd\n",
    "import qgrid\n",
    "import os\n",
    "from IPython.display import display as dp\n",
    "from IPython.display import Markdown as md\n",
    "from IPython.display import IFrame as fr\n",
    "from IPython.display import HTML as ht\n",
    "\n",
    "### Standard Qgrid options\n",
    "grid_options = {\n",
    "    \"fullWidthRows\": True,\n",
    "    \"syncColumnCellResize\": True,\n",
    "    \"forceFitColumns\": False,\n",
    "    \"defaultColumnWidth\": 100,\n",
    "    \"rowHeight\": 23,\n",
    "    \"enableColumnReorder\": True,\n",
    "    \"enableTextSelectionOnCells\": True,\n",
    "    \"editable\": True,\n",
    "    \"autoEdit\": False,\n",
    "    \"explicitInitialization\": True,\n",
    "    \"maxVisibleRows\": 20,\n",
    "    \"minVisibleRows\": 8,\n",
    "    \"sortable\": True,\n",
    "    \"filterable\": True,\n",
    "    \"highlightSelectedCell\": True,\n",
    "    \"highlightSelectedRow\": True,\n",
    "    \"show_toolbar\": False,\n",
    "}\n",
    "\n",
    "goback = \"[Go back to the top](#Jovian-analysis-report)  \"\n",
    "\n",
    "\n",
    "######################################\n",
    "# ACTUAL CONTENT                     #\n",
    "######################################\n",
    "\n",
    "mqc_cell = \"\"\"\n",
    "___\n",
    "## Quality control metrics report  \n",
    "[Open MultiQC graph as a dedicated page by clicking here](results/multiqc.html)  \n",
    "\"\"\"\n",
    "\n",
    "composition_cell = \"\"\"\n",
    "___\n",
    "### Read-based composition of analyzed samples:  \n",
    "[Open the barchart as a dedicated page by clicking here](results/Sample_composition_graph.html)  \n",
    "<br>\n",
    "**Low-quality** reads are those that did not meet the stringency settings as specified in the config file.  \n",
    "**Unclassified** reads are those that could not be assigned to a taxon.  \n",
    "**Remaining** reads are those that could not be assembled into contigs longer than the user specified minimum contig length. \n",
    "\"\"\"\n",
    "\n",
    "krona_cell = \"\"\"\n",
    "___\n",
    "## Metagenomics:\n",
    "___\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "### Interactive metagenomics overview (Krona):\n",
    "[Open Krona graph as a dedicated page by clicking here](results/krona.html)  \n",
    "\"\"\"\n",
    "\n",
    "heatmap_cell_header = \"\"\"\n",
    "### Heatmaps:\n",
    "\"\"\"\n",
    "\n",
    "heatmap_cell_sup = \"\"\"\n",
    "#### Superkingdom heatmap  \n",
    "Open superkingdoms heatmap as a dedicated page by clicking [here](results/heatmaps/Superkingdoms_heatmap.html).  \n",
    "\"\"\"\n",
    "\n",
    "heatmap_cell_vir = \"\"\"\n",
    "#### Virus heatmaps\n",
    "Open virus heatmap as a dedicated page by clicking [here](results/heatmaps/Virus_heatmap.html).  \n",
    "<br>\n",
    "**Please note, many viruses have no \"`order`\" taxonomic rank so always check the \"`family`\" taxonomic rank.**\n",
    "\"\"\"\n",
    "\n",
    "heatmap_cell_pha = \"\"\"\n",
    "#### Phage heatmaps\n",
    "Open phage heatmap as a dedicated page by clicking [here](results/heatmaps/Phage_heatmap.html).  \n",
    "<br>\n",
    "**Please note, many viruses have no \"`order`\" taxonomic rank so always check the \"`family`\" taxonomic rank.**\n",
    "\"\"\"\n",
    "\n",
    "heatmap_cell_bac = \"\"\"\n",
    "#### Bacteria heatmaps\n",
    "Open bacteria heatmap as a dedicated page by clicking [here](results/heatmaps/Bacteria_heatmap.html).  \n",
    "\"\"\"\n",
    "\n",
    "clas_scf_cell = \"\"\"\n",
    "___\n",
    "### Classified scaffolds:\n",
    "\"\"\"\n",
    "\n",
    "unclas_scf_cell = \"\"\"\n",
    "___\n",
    "### Unclassified scaffolds (\"Dark Matter\"):\n",
    "\"\"\"\n",
    "\n",
    "unclas_byNoLCA_cell = \"\"\"\n",
    "___\n",
    "### Ambigious LCA scaffolds:\n",
    "These scaffolds were deemed ambigious by the lowest common ancestor (LCA) analysis, please perform manual inspection. This often occurs for bacteria and (pro)phages.  \n",
    "\"\"\"\n",
    "\n",
    "virhosts_cell = \"\"\"\n",
    "___\n",
    "## Predicted virus hosts:\n",
    "___\n",
    "\"\"\"\n",
    "\n",
    "virtyping_cell = \"\"\"\n",
    "___\n",
    "## Virus typing results:\n",
    "___\n",
    "\"\"\"\n",
    "\n",
    "virtyping_nov_cell = \"\"\"\n",
    "### Norovirus typing tool output:  \n",
    "[Link to the norovirus typing tool](https://www.rivm.nl/mpf/typingtool/norovirus/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_rva_cell = \"\"\"\n",
    "### Rotavirus A typing tool output:  \n",
    "[Link to the Rotavirus A typing tool](https://www.rivm.nl/mpf/typingtool/rotavirusa/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_env_cell = \"\"\"\n",
    "### Enterovirus typing tool output:  \n",
    "[Link to the enterovirus typing tool](https://www.rivm.nl/mpf/typingtool/enterovirus/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_hav_cell = \"\"\"\n",
    "### Hepatitis A typing tool output:  \n",
    "[Link to the hepatatis A typing tool](https://www.rivm.nl/mpf/typingtool/hav/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_hev_cell = \"\"\"\n",
    "### Hepatitis E typing tool output:  \n",
    "[Link to the hepatatis E typing tool](https://www.rivm.nl/mpf/typingtool/hev/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_hpv_cell = \"\"\"\n",
    "### Human Papillomavirus typing tool output:  \n",
    "[Link to the HPV typing tool](https://www.rivm.nl/mpf/typingtool/papillomavirus/)  \n",
    "\"\"\"\n",
    "\n",
    "virtyping_fla_cell = \"\"\"\n",
    "### Flavivirus typing tool output:  \n",
    "[Link to the flavivirus typing tool](https://www.rivm.nl/mpf/typingtool/flavivirus/)  \n",
    "\"\"\"\n",
    "\n",
    "scf_viewer_cell_ILM_META = \"\"\"\n",
    "___\n",
    "## Scaffold viewer:\n",
    "**Containing: SNPs and minority variants (quasispecies), predicted ORFs, depth of coverage graph, GC contents graph**\n",
    "___\n",
    "N.B. Depending on the depth of coverage of the selected contig it can be <b>(very) slow, or downright crash your browser</b>. This is a <b>client-sided</b> problem, meaning, your computer isn't powerful enough.  \n",
    "Open the scaffold viewer as a dedicated page by clicking [here](results/igv.html).  \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "snp_table_cell = \"\"\"\n",
    "___\n",
    "## Minority variant table:\n",
    "___\n",
    "\"\"\"\n",
    "\n",
    "snp_table_empty_cell = \"\"\"\n",
    "Either no SNP's were classified, maybe because you've set the minimum allele-frequency too high? Or something went wrong, please doublecheck the logfiles below:\n",
    "    **logs/SNP_calling_[sample_name].log**\n",
    "    **logs/Concat_filtered_SNPs.log**\n",
    "\"\"\"\n",
    "\n",
    "logging_header_cell = \"\"\"\n",
    "___\n",
    "# Logging and audit-trail: \n",
    "___\n",
    "\"\"\"\n",
    "\n",
    "snakemake_report_cell = \"\"\"\n",
    "### Snakemake summary statistics\n",
    "[Open Snakemake summary statistics as a dedicated page by clicking here](results/snakemake_report.html)\n",
    "\"\"\"\n",
    "\n",
    "logfile_index = \"\"\"\n",
    "<script>\n",
    "function goBack() {\n",
    "    window.history.back()\n",
    "}\n",
    "</script>\n",
    "\n",
    "<button onclick=\"goBack()\">Click this button to go back</button>\n",
    "\n",
    "<div style=\"text-align: center\">\n",
    "    <iframe src=\"results/logfiles_index.html\" width=100% height=980></iframe>\n",
    "</div>\n",
    "\"\"\"\n",
    "\n",
    "acknowledgements_header_cell = \"\"\"\n",
    "___\n",
    "# Acknowledgements:\n",
    "___\n",
    "\"\"\"\n",
    "\n",
    "######################################\n",
    "# WRITE-FUNCTIONS                    #\n",
    "######################################\n",
    "\n",
    "\n",
    "def multiqc():\n",
    "    if os.path.exists(\"results/multiqc.html\"):\n",
    "        dp(md(mqc_cell))\n",
    "        dp(fr(\"results/multiqc.html\", \"100%\", \"980px\"))\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def composition():\n",
    "    if os.path.exists(\"results/Sample_composition_graph.html\"):\n",
    "        dp(md(composition_cell))\n",
    "        dp(fr(\"results/Sample_composition_graph.html\", \"100%\", \"980px\"))\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def krona():\n",
    "    if os.path.exists(\"results/krona.html\"):\n",
    "        dp(md(krona_cell))\n",
    "        dp(fr(\"results/krona.html\", \"100%\", \"980px\"))\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def heatmaps():\n",
    "    if os.path.exists(\"results/heatmaps\"):\n",
    "        dp(md(heatmap_cell_header))\n",
    "        if os.path.exists(\"results/heatmaps/Superkingdoms_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_sup))\n",
    "            dp(fr(\"results/heatmaps/Superkingdoms_heatmap.html\", \"100%\", \"700px\"))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Virus_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_vir))\n",
    "            dp(fr(\"results/heatmaps/Virus_heatmap.html\", \"100%\", \"700px\"))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Phage_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_pha))\n",
    "            dp(fr(\"results/heatmaps/Phage_heatmap.html\", \"100%\", \"700px\"))\n",
    "            dp(md(goback))\n",
    "        if os.path.exists(\"results/heatmaps/Bacteria_heatmap.html\"):\n",
    "            dp(md(heatmap_cell_bac))\n",
    "            dp(fr(\"results/heatmaps/Bacteria_heatmap.html\", \"100%\", \"700px\"))\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "def scaffoldviewer():\n",
    "    if os.path.exists(\"results/igv.html\"):\n",
    "        dp(md(scf_viewer_cell_ILM_META))\n",
    "        dp(fr(\"results/igv.html\", \"100%\", \"980px\"))\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def audit():\n",
    "    if os.path.exists(\"logs/\"):\n",
    "        dp(md(logging_header_cell))\n",
    "        dp(md(\"### Sample sheet\"))\n",
    "        sheet = open(\"results/samplesheet.yaml\", \"r\")\n",
    "        print(sheet.read())\n",
    "        dp(md(snakemake_report_cell))\n",
    "        if os.path.exists(\"results/snakemake_report.html\"):\n",
    "            dp(fr(\"results/snakemake_report.html\", \"100%\", \"980px\"))\n",
    "        else:\n",
    "            print(\n",
    "                \"\"\"\n",
    "We couldn't find a valid snakemake report.\n",
    "This indicates that something went wrong during the used workflow. Please run your analysis again.\n",
    "            \"\"\"\n",
    "            )\n",
    "        dp(md(\"### All log-files:\"))\n",
    "        dp(ht(logfile_index))\n",
    "        dp(md('### Full software list in \"Jovian\" environment:'))\n",
    "        conda_log = open(\"results/log_conda.txt\", \"r\")\n",
    "        print(conda_log.read())\n",
    "        dp(md(\"### Database versions:\"))\n",
    "        db_log = open(\"results/log_db.txt\", \"r\")\n",
    "        print(db_log.read())\n",
    "        dp(md('### Unique methodological \"fingerprint\":'))\n",
    "        id_log = open(\"results/log_git.txt\", \"r\")\n",
    "        print(id_log.read())\n",
    "        dp(md(\"### Snakemake config files:\"))\n",
    "        conf_log = open(\"results/log_config.txt\", \"r\")\n",
    "        print(conf_log.read())\n",
    "        dp(md(goback))\n",
    "\n",
    "\n",
    "def acks():\n",
    "    dp(md(acknowledgements_header_cell))\n",
    "    with open(\"/opt/Jovian/files/acknowledgements.md\", \"r\") as ack_base:\n",
    "        ack_list = ack_base.read()\n",
    "    dp(md(ack_list))\n",
    "    with open(\"/opt/Jovian/files/authors.md\", \"r\") as ath_base:\n",
    "        ath_list = ath_base.read()\n",
    "    dp(md(ath_list))\n",
    "\n",
    "\n",
    "### interactive tables and/or grids\n",
    "class Classified_grid:\n",
    "    def dataframe(clas):\n",
    "        if os.path.exists(\"results/all_taxClassified.tsv\") and os.path.getsize(\"results/all_taxClassified.tsv\") > 0:\n",
    "            clas.df = pd.read_csv(\"results/all_taxClassified.tsv\", sep=\"\\t\")\n",
    "            clas.grid = qgrid.show_grid(clas.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(clas):\n",
    "        clas.dataframe()\n",
    "        if os.path.exists(\"results/all_taxClassified.tsv\") and os.path.getsize(\"results/all_taxClassified.tsv\") > 0:\n",
    "            dp(md(clas_scf_cell))\n",
    "            dp(clas.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class Unclassified_grid:\n",
    "    def dataframe(unclas):\n",
    "        if os.path.exists(\"results/all_taxUnclassified.tsv\") and os.path.getsize(\"results/all_taxUnclassified.tsv\") > 0:\n",
    "            unclas.df = pd.read_csv(\"results/all_taxUnclassified.tsv\", sep=\"\\t\")\n",
    "            unclas.grid = qgrid.show_grid(unclas.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(unclas):\n",
    "        unclas.dataframe()\n",
    "        if os.path.exists(\"results/all_taxUnclassified.tsv\") and os.path.getsize(\"results/all_taxUnclassified.tsv\") > 0:\n",
    "            dp(md(unclas_scf_cell))\n",
    "            dp(unclas.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class Unclassified_byNoLCA_grid:\n",
    "    def dataframe(noLCAclas):\n",
    "        if os.path.exists(\"results/all_noLCA.tsv\") and os.path.getsize(\"results/all_noLCA.tsv\") > 0:\n",
    "            noLCAclas.df = pd.read_csv(\"results/all_noLCA.tsv\", sep=\"\\t\")\n",
    "            noLCAclas.grid = qgrid.show_grid(noLCAclas.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(noLCAclas):\n",
    "        noLCAclas.dataframe()\n",
    "        if os.path.exists(\"results/all_noLCA.tsv\") and os.path.getsize(\"results/all_noLCA.tsv\") > 0:\n",
    "            dp(md(unclas_byNoLCA_cell))\n",
    "            dp(noLCAclas.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class Predict_virhosts:\n",
    "    def dataframe(predicts):\n",
    "        if os.path.exists(\"results/all_virusHost.tsv\") and os.path.getsize(\"results/all_virusHost.tsv\") > 0:\n",
    "            predicts.df = pd.read_csv(\"results/all_virusHost.tsv\", sep=\"\\t\")\n",
    "            predicts.grid = qgrid.show_grid(predicts.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(predicts):\n",
    "        predicts.dataframe()\n",
    "        if os.path.exists(\"results/all_virusHost.tsv\") and os.path.getsize(\"results/all_virusHost.tsv\") > 0:\n",
    "            dp(md(virhosts_cell))\n",
    "            dp(predicts.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "def virtypingheader():\n",
    "    if os.path.exists(\"results/typingtools\"):\n",
    "        dp(md(virtyping_cell))\n",
    "\n",
    "\n",
    "class nov_typing:\n",
    "    def dataframe(nov):\n",
    "        if os.path.exists(\"results/typingtools/all_nov-TT.csv\") and os.path.getsize(\"results/typingtools/all_nov-TT.csv\") > 0:\n",
    "            nov.df = pd.read_csv(\"results/typingtools/all_nov-TT.csv\")\n",
    "            nov.grid = qgrid.show_grid(nov.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(nov):\n",
    "        nov.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_nov-TT.csv\") and os.path.getsize(\"results/typingtools/all_nov-TT.csv\") > 0:\n",
    "            dp(md(virtyping_nov_cell))\n",
    "            dp(nov.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class rva_typing:\n",
    "    def dataframe(rva):\n",
    "        if os.path.exists(\"results/typingtools/all_rva-TT.csv\") and os.path.getsize(\"results/typingtools/all_rva-TT.csv\") > 0:\n",
    "            rva.df = pd.read_csv(\"results/typingtools/all_rva-TT.csv\")\n",
    "            rva.grid = qgrid.show_grid(rva.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(rva):\n",
    "        rva.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_rva-TT.csv\") and os.path.getsize(\"results/typingtools/all_rva-TT.csv\") > 0:\n",
    "            dp(md(virtyping_rva_cell))\n",
    "            dp(rva.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class env_typing:\n",
    "    def dataframe(env):\n",
    "        if os.path.exists(\"results/typingtools/all_ev-TT.csv\") and os.path.getsize(\"results/typingtools/all_ev-TT.csv\") > 0:\n",
    "            env.df = pd.read_csv(\"results/typingtools/all_ev-TT.csv\")\n",
    "            env.grid = qgrid.show_grid(env.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(env):\n",
    "        env.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_ev-TT.csv\") and os.path.getsize(\"results/typingtools/all_ev-TT.csv\") > 0:\n",
    "            dp(md(virtyping_env_cell))\n",
    "            dp(env.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class hav_typing:\n",
    "    def dataframe(hav):\n",
    "        if os.path.exists(\"results/typingtools/all_hav-TT.csv\") and os.path.getsize(\"results/typingtools/all_hav-TT.csv\") > 0:\n",
    "            hav.df = pd.read_csv(\"results/typingtools/all_hav-TT.csv\")\n",
    "            hav.grid = qgrid.show_grid(hav.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(hav):\n",
    "        hav.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_hav-TT.csv\") and os.path.getsize(\"results/typingtools/all_hav-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hav_cell))\n",
    "            dp(hav.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class hev_typing:\n",
    "    def dataframe(hev):\n",
    "        if os.path.exists(\"results/typingtools/all_hev-TT.csv\") and os.path.getsize(\"results/typingtools/all_hev-TT.csv\") > 0:\n",
    "            hev.df = pd.read_csv(\"results/typingtools/all_hev-TT.csv\")\n",
    "            hev.grid = qgrid.show_grid(hev.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(hev):\n",
    "        hev.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_hev-TT.csv\") and os.path.getsize(\"results/typingtools/all_hev-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hev_cell))\n",
    "            dp(hev.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class hpv_typing:\n",
    "    def dataframe(hpv):\n",
    "        if os.path.exists(\"results/typingtools/all_pv-TT.csv\") and os.path.getsize(\"results/typingtools/all_pv-TT.csv\") > 0:\n",
    "            hpv.df = pd.read_csv(\"results/typingtools/all_pv-TT.csv\")\n",
    "            hpv.grid = qgrid.show_grid(hpv.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(hpv):\n",
    "        hpv.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_pv-TT.csv\") and os.path.getsize(\"results/typingtools/all_pv-TT.csv\") > 0:\n",
    "            dp(md(virtyping_hpv_cell))\n",
    "            dp(hpv.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class fla_typing:\n",
    "    def dataframe(fla):\n",
    "        if os.path.exists(\"results/typingtools/all_flavi-TT.csv\") and os.path.getsize(\"results/typingtools/all_flavi-TT.csv\") > 0:\n",
    "            fla.df = pd.read_csv(\"results/typingtools/all_flavi-TT.csv\")\n",
    "            fla.grid = qgrid.show_grid(fla.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(fla):\n",
    "        fla.dataframe()\n",
    "        if os.path.exists(\"results/typingtools/all_flavi-TT.csv\") and os.path.getsize(\"results/typingtools/all_flavi-TT.csv\") > 0:\n",
    "            dp(md(virtyping_fla_cell))\n",
    "            dp(fla.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "class snp_variants:\n",
    "    def dataframe(snps):\n",
    "        if os.path.exists(\"results/all_filtered_SNPs.tsv\") and os.path.getsize(\"results/all_filtered_SNPs.tsv\") > 0:\n",
    "            snps.df = pd.read_csv(\"results/all_filtered_SNPs.tsv\", sep=\"\\t\")\n",
    "            snps.grid = qgrid.show_grid(snps.df, grid_options=grid_options)\n",
    "            return\n",
    "\n",
    "    def draw(snps):\n",
    "        snps.dataframe()\n",
    "        if os.path.exists(\"results/all_filtered_SNPs.tsv\") and os.path.getsize(\"results/all_filtered_SNPs.tsv\") > 0:\n",
    "            dp(md(snp_table_cell))\n",
    "            dp(snps.grid)\n",
    "            dp(md(goback))\n",
    "\n",
    "\n",
    "classified_scaffolds = Classified_grid()\n",
    "unclassified_scaffolds = Unclassified_grid()\n",
    "unclassified_byNoLCA = Unclassified_byNoLCA_grid()\n",
    "predicted_virushosts = Predict_virhosts()\n",
    "virus_typing_nov = nov_typing()\n",
    "virus_typing_rva = rva_typing()\n",
    "virus_typing_env = env_typing()\n",
    "virus_typing_hav = hav_typing()\n",
    "virus_typing_hev = hev_typing()\n",
    "virus_typing_hpv = hpv_typing()\n",
    "virus_typing_fla = fla_typing()\n",
    "minority_variant_table = snp_variants()\n",
    "\n",
    "### actually draw the contents\n",
    "\n",
    "multiqc()\n",
    "composition()\n",
    "krona()\n",
    "heatmaps()\n",
    "\n",
    "classified_scaffolds.draw()\n",
    "unclassified_scaffolds.draw()\n",
    "unclassified_byNoLCA.draw()\n",
    "\n",
    "predicted_virushosts.draw()\n",
    "\n",
    "virtypingheader()\n",
    "virus_typing_nov.draw()\n",
    "virus_typing_rva.draw()\n",
    "virus_typing_env.draw()\n",
    "virus_typing_hav.draw()\n",
    "virus_typing_hev.draw()\n",
    "virus_typing_hpv.draw()\n",
    "virus_typing_fla.draw()\n",
    "\n",
    "scaffoldviewer()\n",
    "\n",
    "minority_variant_table.draw()\n",
    "\n",
    "audit()\n",
    "acks()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "Jovian is available on [GitHub](https://github.com/DennisSchmitz/Jovian) under a [AGPL license](https://www.gnu.org/licenses/agpl-3.0). The virus-typing tools are public services hosted by the [RIVM](https://www.rivm.nl/en) and developed independently of Jovian.\n",
    "___\n",
    "*This study was financed under European Union’s Horizon H2020 grants COMPARE and VEO (grant no. 643476 and 874735).*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dp(md(goback))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deleteable": false,
    "init_cell": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%javascript\n",
    "$('<div id=\"toc\"></div>').css({position: 'fixed', top: '120px', left: 0}).appendTo(document.body);\n",
    "$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js');"
   ]
  }
 ],
 "metadata": {
  "hide_input": true,
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "210px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "position": {
    "height": "413px",
    "left": "1476px",
    "right": "20px",
    "top": "120px",
    "width": "350px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
